nsxt/resource_nsxt_node_user.go

```diff
@@ -0,0 +1,238 @@
+				ValidateFunc: validation.IntBetween(0, 9999),
+			},
+			"password_change_warning": {
+				Type:         schema.TypeInt,
+				Description:  "Number of days before user receives warning message of password expiration",
+				Optional:     true,
+				Default:      7,
+				ValidateFunc: validation.IntBetween(0, 9999),
+			},
+			"password_reset_required": {
+				Type:        schema.TypeBool,
+				Description: "Boolean value that states if a password reset is required\t",
+				Computed:    true,
+			},
+			"status": {
+				Type:         schema.TypeString,
+				Description:  "User status",
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice(nodeUserStatus, false),
+				Computed:     true,
+			},
+			"userid": {
+				Type:        schema.TypeInt,
+				Description: "Numeric id for the user",
+				Computed:    true,
+			},
+			"username": {
+				Type:         schema.TypeString,
+				Description:  "User login name",
+				ValidateFunc: validateUsername(),
+				Required:     true,
+			},
+		},
+	}
+}
+
+func validateUsername() schema.SchemaValidateFunc {
+	return func(i interface{}, k string) (s []string, es []error) {
+		v, ok := i.(string)
+		if !ok {
+			es = append(es, fmt.Errorf("expected type of %s to be string", k))
+			return
+		}
+
+		r := regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9@-_.\\-]*$")
+		if ok := r.MatchString(v); !ok {
+			es = append(es, fmt.Errorf("must be a valid username matching: ^[a-zA-Z][a-zA-Z0-9@-_.\\-]*$"))
+			return
+		}
+
+		if len(v) < 1 || len(v) > 32 {
+			es = append(es, fmt.Errorf("expected length of %s to be in the range (%d - %d), got %s", k, 1, 32, v))
+			return
+		}
+		return
+	}
+}
+
+func resourceNsxtNodeUserCreate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := node.NewUsersClient(connector)
+
+	fullName := d.Get("full_name").(string)
+	passwordChangeFrequency := int64(d.Get("password_change_frequency").(int))
+	passwordChangeWarning := int64(d.Get("password_change_warning").(int))
+	username := d.Get("username").(string)
+	password := d.Get("password").(string)
+
+	userProp := nsxModel.NodeUserProperties{
+		FullName:                &fullName,
+		PasswordChangeFrequency: &passwordChangeFrequency,
+		PasswordChangeWarning:   &passwordChangeWarning,
+		Username:                &username,
+	}
+	if len(password) > 0 {
+		userProp.Password = &password
+	}
+
+	user, err := client.Createuser(userProp)
+	if err != nil {
+		return fmt.Errorf("failed to create Nsxt Node user: %s", err.Error())
+	}
+	d.Set("userid", user.Userid)
+	d.SetId(strconv.Itoa(int(*user.Userid)))
+
+	return resourceNsxtNodeUserRead(d, m)
+}
+
+func resourceNsxtNodeUserRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := node.NewUsersClient(connector)
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("error obtaining logical object id")
+	}
+	user, err := client.Get(id)
+	if err != nil {
+		if isNotFoundError(err) {
```

Comment:
This behavior is covered with `handleReadError`

Reply:
done